framework/src/base/NonlinearSystem.C

```diff
@@ -149,7 +183,25 @@ NonlinearSystem::solve()
   setInitialSolution();
 
   if (_use_finite_differenced_preconditioner)
+  {
+    _transient_sys.nonlinear_solver->fd_residual_object = &_fd_residual_functor;
     setupFiniteDifferencedPreconditioner();
+  }
+
+#ifdef LIBMESH_HAVE_PETSC
+  PetscNonlinearSolver<Real> & solver =
+      static_cast<PetscNonlinearSolver<Real> &>(*_transient_sys.nonlinear_solver);
+  Moose::SolveType moose_solve_type = _fe_problem.solverParams()._type;
+  if (moose_solve_type == Moose::ST_PJFNK || moose_solve_type == Moose::ST_JFNK)
+  {
+    if (moose_solve_type == Moose::ST_PJFNK)
+      solver.solve_type() = PetscNonlinearSolver<Real>::MF_OPERATOR;
+    else
+      solver.solve_type() = PetscNonlinearSolver<Real>::MF;
+
+    solver.mffd_residual_object = &_fd_residual_functor;
```

Comment:
Is it better to create a `_mffd_residual_functor`? For FDColoring, the residual evaluation could be called several times for a given linear iteration while it is called once for MFFD.

Reply:
This is moot now with the form of the Petsc API changes